native/game_backend/src/space_hash_grid.rs

```diff
@@ -0,0 +1,152 @@
+use crate::{map::Position, player::Player, projectile::{Projectile, self}};
+use rustler::NifMap;
+use std::collections::HashMap;
+
+#[derive(Clone, Debug, rustler::NifMap)]
+pub struct Vec2 {
+    pub x: f32,
+    pub y: f32,
+}
+
+impl From<Position> for Vec2 {
+    fn from(value: Position) -> Self {
+        let x = value.x as f32;
+        let y = value.y as f32;
+        Self { x, y }
+    }
+}
+
+impl From<Vec2> for Position {
+    fn from(value: Vec2) -> Self {
+        let x = value.x as i64;
+        let y = value.y as i64;
+        Self { x, y }
+    }
+}
+
+pub type Bucket = Vec<GameEntity>;
+
+#[derive(Clone, Debug, rustler::NifMap)]
+pub struct GameEntity {
+    pub position: Vec2,
+    pub radius: f32,
+    pub is_blocking: bool
+}
+
+impl From<&Player> for GameEntity {
+    fn from(player: &Player) -> GameEntity {
+        let position = player.position.into();
+        let radius = player.size as f32;
+        let id = player.id;
+        GameEntity {
+            position,
+            radius,
+            is_blocking: false
+        }
+    }
+}
+
+impl From<&Projectile> for GameEntity {
+    fn from(projectile: &Projectile) -> GameEntity {
+        let position = Vec2::from(projectile.position);
+        GameEntity {
+            radius: projectile.size as f32,
+            position,
+            is_blocking: false
+        }
+    }
+}
+
+#[derive(NifMap)]
+pub struct SpatialHashGrid {
+    cols: u64,
+    rows: u64,
+    bucket_total: u64,
+    buckets: HashMap<u64, Bucket>,
+    sceneheight: u64,
+    cellsize: u64,
+    scenewidth: u64,
+}
+
+impl SpatialHashGrid {
+    pub fn new(scenewidth: u64, sceneheight: u64, cellsize: u64) -> Self {
+        let cols = scenewidth / cellsize;
+        let rows = sceneheight / cellsize;
+        let bucket_total = cols*rows;
+        let mut buckets = HashMap::new();
+        for i in 0..bucket_total {
+            buckets.insert(i, vec![]);
+        }
+        SpatialHashGrid {
+            cols,
+            rows,
+            bucket_total,
+            buckets,
+            sceneheight,
+            cellsize,
+            scenewidth,
+        }
+    }
+    pub fn clear_buckets(&mut self) {
+        self.buckets = HashMap::new();
+        for i in 0..self.bucket_total {
+            self.buckets.insert(i, vec![]);
+        }
```

Comment:
```suggestion
        self.buckets.values_mut().for_each(|b| b.clear());
```
Calling `clear` on each value saves reallocations later, as it reuses the old buffer.

Reply:
Alternatively, I think the following should work too:
```suggestion
        self.buckets = HashMap::new();
        for i in 0..self.bucket_total {
            self.buckets.values_mut().for_each(HashMap::clear);
        }
```